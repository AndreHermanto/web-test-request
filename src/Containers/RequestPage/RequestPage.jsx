import React, {Component} from 'react';
import {
  FormGroup,
  FormControl,
  ControlLabel
} from 'react-bootstrap';


/**
 * RequestPage - this generates the main page sending the test request.
 */
class RequestPage extends Component {
  constructor(props) {
    super(props); 
    
    this.state = {
      testList: []
    };
  }
  
  componentWillMount() {
    this.getList('#','testList'); // # is only temporary url
  }
  
  /**
   * A shared function to retrieve the option list from the gateway
   * e.g., retrieving the test list and gene list
   *
   * @param {String} url - url of the api call
   * @param {String} stateChild - name of the state that the list applies
   */
  getList(url, stateChild) {
    var func = this;
    fetch(url)
      .then(
        (response) => response.text()
      )
      .then(
        (json) => {
          func.setState(
            (state) => {
              state[stateChild] = json;
              return state;
            }
          );
        }
      );
  }
  
  /**
   * A shared function to generate options from a list contained in state.
   * e.g., generating options for the test list and gene list
   *
   * @param {String} stateChild - name of the state that the list applies
   */
  genOptions(stateChild) {
    this.state[stateChild].map(
      (item) => {
        return <option value={item}>{item}</option>
      }
    );
  }
  
  render() {
    return (
      
      <div>
        <h4>Web Test Request</h4>

        <FormGroup>
          <ControlLabel>Available Test List</ControlLabel>
          <FormControl 
            componentClass="select" 
            placeholder="Select test" 
            name="test"
          >
            
          {this.genOptions('testList')}
            
          </FormControl>
        </FormGroup>
      </div>
      
    );
  }
};

export default RequestPage;