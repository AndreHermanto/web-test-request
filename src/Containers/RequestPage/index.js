import React, {Component} from 'react';
import {
  FormGroup,
  FormControl,
  ControlLabel
} from 'react-bootstrap';

/**
 * RequestPage - this generates the main page sending the test request.
 */
function RequestPage(props) {
  return (

    <div>
      <h4>Web Test Request</h4>
      <FormGroup>
        <ControlLabel>Available Test List</ControlLabel>
        <FormControl 
          componentClass="select" 
          placeholder="Select test" 
          name="test"
        >

        {
          props.testList.map(
            (test) => {
              return <option key={test} value={test}>{test}</option>
            }
          )
        }

        </FormControl>
      </FormGroup>
    </div>

  );
}

export default RequestPage;