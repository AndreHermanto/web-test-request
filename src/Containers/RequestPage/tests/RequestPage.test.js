import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import RequestPage from './../index';

describe('RequestPage', () => {
  it('renders without crashing - initial state with empty geneList', () => {
    const tree = renderer.create(
      <RequestPage 
        testList={['test1','test2','test3']}
        geneList={[]}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it('renders without crashing - after a test is selected, geneList defined', () => {
    const tree = renderer.create(
      <RequestPage 
        testList={['test1','test2','test3']}
        geneList={[
          {
            "disorder": "xxx1",
            "genes": ["gene1", "gene2"]	
          }, 
          {
            "disorder": "xxx2",
            "genes": ["gene1", "gene2"]	
          }]
        }
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
