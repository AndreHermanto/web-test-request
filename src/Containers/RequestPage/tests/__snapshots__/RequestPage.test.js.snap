exports[`RequestPage renders without crashing - after a test is selected, geneList defined 1`] = `
<div>
  <h4>
    Web Test Request
  </h4>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Name
    </label>
    <input
      className="form-control"
      id={undefined}
      name="name"
      onChange={undefined}
      placeholder="Hello World"
      type="text" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Available Test List
    </label>
    <select
      className="form-control"
      id={undefined}
      name="test"
      onChange={undefined}
      placeholder="Select test"
      type={undefined}>
      <option
        value="test1">
        test1
      </option>
      <option
        value="test2">
        test2
      </option>
      <option
        value="test3">
        test3
      </option>
    </select>
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Gene List
    </label>
    <select
      className="form-control"
      id={undefined}
      name="geneList"
      placeholder="Select gene list"
      type={undefined}>
      <option
        value="xxx1">
        xxx1
      </option>
      <option
        value="xxx2">
        xxx2
      </option>
    </select>
  </div>
  <button
    className="btn btn-success"
    disabled={false}
    onClick={undefined}
    type="submit">
    Submit Request
  </button>
</div>
`;

exports[`RequestPage renders without crashing - initial state with empty geneList 1`] = `
<div>
  <h4>
    Web Test Request
  </h4>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Name
    </label>
    <input
      className="form-control"
      id={undefined}
      name="name"
      onChange={undefined}
      placeholder="Hello World"
      type="text" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Available Test List
    </label>
    <select
      className="form-control"
      id={undefined}
      name="test"
      onChange={undefined}
      placeholder="Select test"
      type={undefined}>
      <option
        value="test1">
        test1
      </option>
      <option
        value="test2">
        test2
      </option>
      <option
        value="test3">
        test3
      </option>
    </select>
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Gene List
    </label>
    <select
      className="form-control"
      id={undefined}
      name="geneList"
      placeholder="Select gene list"
      type={undefined} />
  </div>
  <button
    className="btn btn-success"
    disabled={false}
    onClick={undefined}
    type="submit">
    Submit Request
  </button>
</div>
`;
